feat(windows): Keyman Core Integration - Support for IMX DLLs 🥑

rc-swag opened the pull request:
Fixes #5650

rc-swag commented on common/core/desktop/include/keyman/keyboardprocessor.h:
```diff
@@ -131,6 +131,10 @@ typedef struct km_kbp_options     km_kbp_options;
 
+// Callback functions
+//
+typedef KMN_API uint8_t (*km_kbp_keyboard_imx_platform)(km_kbp_state*, uint32_t, void*);
```
Callback function pointer declared here.

rc-swag commented on common/core/desktop/tests/unit/kmx/kmx-imx.cpp:
```diff
@@ -0,0 +1,246 @@
+
+
+extern "C"
+{
```
This is the callback used in a unit test that runs without error.

rc-swag commented on windows/src/engine/keyman32/calldll.h:
```diff
@@ -24,9 +24,19 @@ BOOL UnloadDLLs(LPINTKEYBOARDINFO lpkbi);
 BOOL IsIMWindow(HWND hwnd);
 
 void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum);
+extern "C" uint8_t IM_CallBackCore(km_kbp_state *km_state, uint32_t UniqueStoreNo, void *callbackObject);
```
This is the windows engine callback function declaration.

rc-swag commented:
An outstanding issue is if an IM window is up waiting for a user code point selection and other keys are pressed on the keyboard those characters will be added to the context and output before the select character from the IMC window.  This behaviour is the same as the current windows implementation.  However, the core implementation the character entered is output 3 times. 
For example with IM Test Loaded Open notepad
Type `d` then `a` which will then wait for input from the IMC window, before selecting a character press `t`.
Then select "æ"
The output in the app will be: `dtttæ`

Looking at the logging I can see that 
```if (fOutputKeystroke && !_td->app->IsQueueEmpty()) {``` in `kmprocess.cpp` `ProcessHook` is evaluating to true and the
key gets synthesised twice. The action to output the char is already in the queue (which is why the if evaluates to true). Therefore the key is added 3 times.

rc-swag commented:
# 🚧User Testing

* **TEST_IMSAMPLE_KEYBOARD:**


* **TEST_SIMPLIFIED_CHINESE :**

* **TEST_OTHER_IMXDLL:** 
Keyboards

mcdurdin commented:
The linux build has four failing tests:

```
[Step 3/4] FAIL: run-tests.sh 38 /integration-tests/surrounding-text/k_041___long_context_and_deadkeys
[Step 3/4] FAIL: run-tests.sh 39 /integration-tests/surrounding-text/k_042___long_context_and_split_deadkeys
[Step 3/4] FAIL: run-tests.sh 83 /integration-tests/no-surrounding-text/k_041___long_context_and_deadkeys
[Step 3/4] FAIL: run-tests.sh 84 /integration-tests/no-surrounding-text/k_042___long_context_and_split_deadkeys
```

I think we need to figure out why that is before reviewing?

mcdurdin commented:
Also, can you avoid including the recompiled .kmx files unless they have meaningful changes? It makes the PR noisy!

rc-swag commented:
> The linux build has four failing tests:
> 
> ```
> [Step 3/4] FAIL: run-tests.sh 38 /integration-tests/surrounding-text/k_041___long_context_and_deadkeys
> [Step 3/4] FAIL: run-tests.sh 39 /integration-tests/surrounding-text/k_042___long_context_and_split_deadkeys
> [Step 3/4] FAIL: run-tests.sh 83 /integration-tests/no-surrounding-text/k_041___long_context_and_deadkeys
> [Step 3/4] FAIL: run-tests.sh 84 /integration-tests/no-surrounding-text/k_042___long_context_and_split_deadkeys
> ```
> 
> I think we need to figure out why that is before reviewing?

Moving back to draft why I sort this out thanks @ermshiperete  for point me in the right direction

rc-swag commented:
[IMSAMPLE.zip](https://github.com/keymanapp/keyman/files/7878812/IMSAMPLE.zip)
Attached is the built imsample keyboard found at [imsample source](https://github.com/keymanapp/keyman/tree/master/windows/src/developer/samples/imsample)

rc-swag commented:
> > The linux build has four failing tests:
> > ```
> > [Step 3/4] FAIL: run-tests.sh 38 /integration-tests/surrounding-text/k_041___long_context_and_deadkeys
> > [Step 3/4] FAIL: run-tests.sh 39 /integration-tests/surrounding-text/k_042___long_context_and_split_deadkeys
> > [Step 3/4] FAIL: run-tests.sh 83 /integration-tests/no-surrounding-text/k_041___long_context_and_deadkeys
> > [Step 3/4] FAIL: run-tests.sh 84 /integration-tests/no-surrounding-text/k_042___long_context_and_split_deadkeys
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think we need to figure out why that is before reviewing?
> 
> Moving back to draft why I sort this out thanks @ermshiperete for pointing me in the right direction

Reverted all the kmx files that were inadvertently checked in on a commit.

rc-swag commented:
> Also, can you avoid including the recompiled .kmx files unless they have meaningful changes? It makes the PR noisy!

yes done.

mcdurdin commented:
> Moving back to draft why I sort this out thanks @ermshiperete for pointing me in the right direction

What was the answer here? Was it something to do with the rebuild .kmx files?

rc-swag commented:
> at was the answer here? Was it something to do with the rebuild .kmx files?

Yes, I had a commit that was mainly just all code comments but I had added the *.kmx file in that commit. I am guessing they had been built at the time those tests were been worked on if you remember a few weeks back. Since then I have rebased master so the source files are correct. The trick is that on all the other build agents and on my machine it was rebuilding the kmx from the source so it didn't fail. @ermshiperete did force check using the built kmx file in this PR to verify it failed locally.

I went through the commit history and found I only had the one commit of kmx files as mentioned above it was just code comment commit and inadvertently checked in the kmx files.  The cleanest way to fix this was 

 ` for i in 0*.kmx; do git checkout 993655d~1 "$i"; done;` 
 This checked out all test kmx built files to the version before the offending commit.

mcdurdin commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+        actionItems[idx].backspace.expected_type  = KM_KBP_BT_CHAR;
+        actionItems[idx].backspace.expected_value = (DWORD)Uni_SurrogateToUTF32(*CodeUnitPtr, *(CodeUnitPtr + 1));
+      } else {
+        CodeUnitPtr = context.BufMax(SingleCharLength);
+        actionItems[idx].backspace.expected_type  = KM_KBP_BT_CHAR;
+        actionItems[idx].backspace.expected_value = (DWORD)*CodeUnitPtr;
```
What about the case where we are deleting off the start of the context (i.e. context is empty)?

mcdurdin commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+      return FALSE;
+    }
+    km_kbp_context_item *citems = nullptr;
+      if (KM_KBP_STATUS_OK !=
+        (km_kbp_status_codes)kbp_state_get_intermediate_context(_td->lpActiveKeyboard->lpCoreKeyboardState, &citems)){
+          km_kbp_context_items_dispose(citems);
+          return FALSE;
```
Please fix indentation

mcdurdin commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -327,46 +588,53 @@ extern "C" BOOL _declspec(dllexport) WINAPI KMDisplayIM(HWND hwnd, BOOL FShowAlw
 extern "C" BOOL _declspec(dllexport) WINAPI KMHideIM()
 {
-	if(*Globals::hwndIM() != 0)
+  SendDebugMessageFormat(0, sdmKeyboard, 0, "KMHideIM: Enter");
```
Let's comment out these debug messages for now -- no need to log every function call

rc-swag replied:
Yes in the commit history I had a commit where I removed a heap of logging. then when I had the emit keystroke issue with the IMX synthesize key up/key down. I had to put it all back in so I didn't want to remove it all again.
It is a shame we don't have levels of logging like debug, warn, error. ( I think info is usually one level too many).  Seeing any code commented out still gives me shivers as it was drummed into me so hard illegal in SIL (Safety Integrity Level) code.  --- doesn't mean I sometimes forget though. I guess I will just remove it.

mcdurdin commented on windows/src/engine/keyman32/preservedkeymap.cpp:
```diff
@@ -392,6 +392,7 @@ PreservedKeyMap::MapKeyboardCore(km_kbp_keyboard *pKeyboard, PreservedKey **pPre
   }
 
   km_kbp_keyboard_key *key_rule_it = kb_key_list;
+
```
Can we rollback this change and the whitespace changes in kmprocessactions.cpp?

mcdurdin reviewed (changes requested):
Overall, this looks very nice. I have a bunch of formatting comments, sorry.

There is one question in my mind around when the IMX callback is made -- shouldn't it be made in the `!TIPFUpdateable` phase?

rc-swag commented:
> Overall, this looks very nice. I have a bunch of formatting comments, sorry.
> 
> There is one question in my mind around when the IMX callback is made -- shouldn't it be made in the `!TIPFUpdateable` phase?
Yes  see this WIP 
https://github.com/keymanapp/keyman/pull/6122

Yes so we discussed this the other day if you do it the other way around you will have the wrong actions (they will be lost) on the Updatable parse when the core process the keystroke on the Updatealble pass. So we will process everything in the !Updateable Parse. Then use the cached actions in the core on the second parse.

rc-swag commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+    return FALSE;
+  SendDebugMessageFormat(
+      0, sdmKeyboard, 0, "IM_CallBackCore: td loadeded");
+  if (_td->TIPFUpdateable) {  // I4452
+    SendDebugMessageFormat(0, sdmKeyboard, 0, "IM_CallBackCore: td TIPFUpdatable about to call function [%s]", imdh->name);
+    LogContext(_td->lpActiveKeyboard->lpCoreKeyboardState, CONTEXT_CORE);
+    LogContext(_td->lpActiveKeyboard->lpCoreKeyboardState, CONTEXT_INT);
+    (*imdh->function)(_td->state.msg.hwnd, _td->state.vkey, _td->state.charCode, Globals::get_ShiftState());
+  }
```
Yes - seperate PR WIP see https://github.com/keymanapp/keyman/pull/6122

rc-swag commented on common/core/desktop/include/keyman/keyboardprocessor.h:
```diff
@@ -459,6 +464,23 @@ km_kbp_context_shrink(km_kbp_context *context,
+##### Description:
+Return the length of a terminated `km_kbp_context_item` array.
+##### Return:
+The number of items in the list or 0 if `context_items` is null.
```
No - I made it consistent with `km_kbp_options_list_size`

mcdurdin replied:
```suggestion
The number of items in the list, not including terminating item,
or 0 if `context_items` is null.
```

(Could also tweak the description of `km_kbp_options_list_size` accordingly)

rc-swag commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +106,107 @@ kmx_processor::update_option(
+  case KM_KBP_IT_END:
+    // error should not queue empty item
+    return false;
+    break;
```
Not sure of the intent of the question.
There is nothing to stop a developer calling queue_action() with `item->type == KM_KBP_IT_END`, it would be silly but nothing to stop it so the code is reachable. Currently it is protected by the for loop in km_kbp_state_queue_action_items.

mcdurdin replied:
Just `break;` immediately after `return false;` was all I was asking :grin:

rc-swag commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +106,107 @@ kmx_processor::update_option(
+
+    if(!p_last_item_context || *p_last_item_context == 0)   {
+      _kmx.GetActions()->QueueAction(QIT_INVALIDATECONTEXT, 0);
+      // this method(queue_action) could be called outside the processing of a key
+      // stroke, so it cannot set emit keystroke. Synthasize instead;
+      // Currently    QIT_VKEYDOWN QIT_VKEYUP QIT_VSHIFTDOWN UP are not implmented
+      //in the core KM_KBP actions therefore we can not synthasize.
+      // For this Backspace case the Core is going to have to relax the
+      // no QIT_BACK on a empty context. for now.
+      //_kmx.GetActions()->QueueAction(QIT_VKEYDOWN, KM_KBP_VKEY_BKSP);
+      //_kmx.GetActions()->QueueAction(QIT_VKEYUP, KM_KBP_VKEY_BKSP);
+
+      // if we can synthasize backspace remove this QueAction back
```
The core was asserting if the context was empty. Options where to synthesize a keystroke or the core to relax that restriction. As you have commented below that is the way to go. I will remove all these comments.

rc-swag commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +106,107 @@ kmx_processor::update_option(
+  // If the Virtual Key is VK_SPACE and the internal kmx processor has actions
+  // then process that and return. These actions must have been added externally
+  // via the queue_action method.
+  bool has_internal_actions = ((vk == VK_SPACE) && (!_kmx.GetActions()->IsQueueEmpty()));
```
That was your suggestion as I understood it.

mcdurdin replied:
Sorry ... I had forgotten -- this looks good.

rc-swag commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+  if (!Globals::get_CoreIntegration()) {  // TODO: 5442 Remove If and fix indent
+    while (backlen-- > 0)
+      _td->app->QueueAction(QIT_BACK, 0);
+    while (*buf)
+      _td->app->QueueAction(QIT_CHAR, *buf++);
```
So the new code handles surrogates are you saying you want me to updated the old code to handle it as well?

mcdurdin replied:
Ah yes, this is in the old code path; I am comfortable with just leaving it.

rc-swag commented on windows/src/engine/keyman32/calldll.cpp:
```diff
@@ -253,48 +347,215 @@ void CallDLL(LPINTKEYBOARDINFO lpkbi, DWORD storenum)
+      km_kbp_context_items_dispose(citems);
+      return FALSE;
+    }
+    WCHAR *contextString           = new WCHAR[MAXCONTEXT];  // *3 if every context item was a deadkey
```
Good pick up don't know what I was thinking.

rc-swag commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -141,7 +236,10 @@ kmx_processor::process_event(
       switch (a.dwData) {
       case BK_DEFAULT:
         // This only happens if we know we have context to delete. Last item must be a character
-        assert(!state->context().empty());
+
+        // TODO: #5060 If we added a new action to core actions then we
+        // can assert on backspace for a empty context.
+        //assert(!state->context().empty());
```
I don't believe in the IMX callback you will ever hit the `KM_KBP_BT_UNKNOWN` case so the comment can stay as is.

rc-swag commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -240,3 +349,54 @@ km_kbp_keyboard_key * kmx_processor::get_key_list() const  {
+			PKMX_CHAR full_fn_name = wstrtostr(p_store->dpString), lib_name, fn_name;
+
+			lib_name = strtok(full_fn_name, ":");
+			fn_name = strtok(NULL, ":");
```
Yes u16tok made the code even simpler, however I should note that in the windows engine makes the same error of convering to unicode to codepage should we update that as well?

rc-swag commented:
[imsample.zip](https://github.com/keymanapp/keyman/files/7894105/imsample.zip)

mcdurdin commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +103,99 @@ kmx_processor::update_option(
+  DebugLog("Action type is [%d].\n", action_item->type);
+  switch (action_item->type) {
+  case KM_KBP_IT_END:
+    // error should not queue empty item
```
Let's throw a debug log statement here given it's not a thing that should be done and probably indicates a logic bug somewhere...

mcdurdin commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +103,99 @@ kmx_processor::update_option(
+  case KM_KBP_IT_PERSIST_OPT:
+  case KM_KBP_IT_EMIT_KEYSTROKE:
+  case KM_KBP_IT_CAPSLOCK:
+    // Not implemented TODO log message?
```
Yes, let's DebugLog here.

mcdurdin commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -75,43 +103,99 @@ kmx_processor::update_option(
-        ctxt += (km_kbp_cp)c->character;
+  // If the Virtual Key is VK_SPACE and the internal kmx processor has actions
+  // then process that and return. Why - because these actions must have been added externally
+  // via the queue_action method. The IMX extension requests processing of action queue
```
```suggestion
  // via the queue_action method. The client can request processing of action queue
```

We should ensure that this little nuance is documented in the keyboardprocessor.h in `queue_action` and `process_event` comments -- it's an implementation detail that should be made consistent in each of the processors.

---

Alternatively... we could add a tiny little function `km_kbp_process_queued_actions(state)` which avoids this, and refactor `kmx_processor::process_event` and the new `kmx_processor::process_queued_actions` to both call `kmx_processor::internal_process_queued_actions` accordingly? A little bit more work but a whole lot prettier I think. Something like:

```
kmx_processor::process_queued_actions() {
  state->actions().clear();
  return internal_process_queued_actions();
}
```

rc-swag replied:
Oh yes this is great. After our initial discussion I saw didn't need to actually synthesize the keystoke using sendinput or whatever -- it would have caused more issues.  I could just send VK_SPACE straight into process_event I didn't make the next logical step that you have here. This is much nicer for the next developer reading the code.

mcdurdin commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -240,3 +329,45 @@ km_kbp_keyboard_key * kmx_processor::get_key_list() const  {
+
+  km_kbp_keyboard_imx *imx_list = new km_kbp_keyboard_imx[fn_count + 1];
+
+  for(uint32_t i = 0; i < store_cnt; i++) {
+    LPSTORE p_store = &store_array[i];
+    if(p_store->dwSystemID == TSS_CALLDEFINITION) {
+			/* Break the store string into components */
+
+			PKMX_WCHAR full_fn_name = p_store->dpString;
+      PKMX_WCHAR pt = NULL;
+      PKMX_WCHAR lib_name = u16tok(full_fn_name, u':', &pt);
+			PKMX_WCHAR fn_name = u16tok(NULL, u':', &pt);
+
+			if(!lib_name || !fn_name){
+				continue;
+			}
+
+      imx_list[fn_idx].library_name = lib_name;
+      imx_list[fn_idx].function_name = fn_name;
+      imx_list[fn_idx].imx_id = i;
+      fn_idx++;
+		}
+  }
```
`u16tok` is good. Eliminates a memory leak too, but introduces a little problem. This is going to fail if we call `get_imx_list` more than once. In the original Windows core, the underlying store `p_store->dpString` is not mutated. Now, as the string is copied, that doesn't hold here.

(The mutation to the structure that does occur in the old Windows core is to mark call definitions that are invalid or where the load fails, as `TSS_CALLDEFINITION_LOADFAILED`, which avoids trying to load them again.)

We probably need to copy the string, and then free it for each entry, in `km_kbp_keyboard_imx_list_dispose` (which means more internal plumbing and perhaps a bit of internal knowledge crossing the core/kmx keyboardprocessor boundary, ugh). Or something like that?

mcdurdin commented on windows/src/engine/keyman32/appint/appint.h:
```diff
@@ -216,10 +225,24 @@ class AppIntegration:public AppActionQueue
+ * Convert km_kbp_context_item array into an kmx char buffer.
+ * Caller is responsible for freeing the memory.
+ * The length is restricted to a maximum of MAXCONTEXT length. If the number
+ * of input km_kbp_context_items exceeds this length the charaters furtherest
```
```suggestion
 * of input km_kbp_context_items exceeds this length the characters furthest
```

mcdurdin commented on windows/src/engine/keyman32/appint/appint.h:
```diff
@@ -216,10 +225,24 @@ class AppIntegration:public AppActionQueue
+ * from the caret will be truncated.
+ *
+ * @param  contextItems  the input core context array. (km_kbp_context_item)
+ * @param  [out] outBuf  the kmx charater array output. caller to free memory.
```
```suggestion
 * @param  [out] outBuf  the kmx character array output. caller to free memory.
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+   * @param callback_object
+   * @return uint8_t
+   */
+  uint8_t test_imx_callback(km_kbp_state *state, uint32_t imx_id, void *callback_object){
```
```suggestion
uint8_t test_imx_callback(km_kbp_state *state, uint32_t imx_id, void *callback_object) {
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+    a_items[2].type   = KM_KBP_IT_END;
+    km_kbp_state_queue_action_items(state, a_items);
+    break;
+    case 5:
```
```suggestion
  case 5:
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+    a_items[3].type   = KM_KBP_IT_END;
+    km_kbp_state_queue_action_items(state, a_items);
+    break;
+    case 6:
```
```suggestion
  case 6:
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+    a_items[2].type   = KM_KBP_IT_END;
+    km_kbp_state_queue_action_items(state, a_items);
+    break;
+    case 7: {
```
```suggestion
  case 7: {
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+  km_kbp_context_items_dispose(exit_context);
+  delete[] tmp_buf;
+  return 1;
+  }
```
```suggestion
}
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+  delete[] tmp_buf;
+  return 1;
+  }
+};
```
```suggestion
}; // extern "C"
```

mcdurdin commented on common/core/desktop/tests/unit/kmx/kmx_imx.cpp:
```diff
@@ -0,0 +1,269 @@
+// the kmx processor will add this items to its action queue.
+// Finally when the `process_event` call returns we verify the action
+// queue is as expected.
+void test_queue_actions (const km::kbp::path &source_keybard) {
```
```suggestion
void test_queue_actions (const km::kbp::path &source_keyboard) {
```

mcdurdin reviewed (changes requested):
I think this is nearly ready. One fix around `u16tok` has opened up a memory management issue...

> > There is one question in my mind around when the IMX callback is made -- shouldn't it be made in the `!TIPFUpdateable` phase?
> 
> Yes it will - see this WIP #6122

Gotcha, didn't realise that you had those changes in a separate PR. Sorry for the confusion. Re #6122, should that be based on this PR and not directly on master?

mcdurdin commented on common/core/desktop/src/kmx/kmx_processor.cpp:
```diff
@@ -240,3 +329,45 @@ km_kbp_keyboard_key * kmx_processor::get_key_list() const  {
+    if(p_store->dwSystemID == TSS_CALLDEFINITION) {
+			/* Break the store string into components */
+
+			PKMX_WCHAR full_fn_name = p_store->dpString;
+      PKMX_WCHAR pt = NULL;
+      PKMX_WCHAR lib_name = u16tok(full_fn_name, u':', &pt);
+			PKMX_WCHAR fn_name = u16tok(NULL, u':', &pt);
+
+			if(!lib_name || !fn_name){
+				continue;
+			}
+
+      imx_list[fn_idx].library_name = lib_name;
+      imx_list[fn_idx].function_name = fn_name;
+      imx_list[fn_idx].imx_id = i;
+      fn_idx++;
```
```suggestion
			PKMX_WCHAR full_fn_name = u16dup(p_store->dpString);
      PKMX_WCHAR pt = NULL;
      PKMX_WCHAR lib_name = u16tok(full_fn_name, u':', &pt);
			PKMX_WCHAR fn_name = u16tok(NULL, u':', &pt);

			if(!lib_name || !fn_name){
				continue;
			}

      imx_list[fn_idx].library_name = lib_name;
      imx_list[fn_idx].function_name = u16dup(fn_name);
      imx_list[fn_idx].imx_id = i;
      fn_idx++;
```

And in the dispose:

```
delete [] library_name; // from u16dup
delete [] function_name; // from u16dup
```

mcdurdin commented on common/core/desktop/src/kmx/kmx_xstring.cpp:
```diff
@@ -103,6 +103,12 @@ km_kbp_cp *km::kbp::kmx::u16tok(km_kbp_cp *p, km_kbp_cp ch, km_kbp_cp **ctx) {
 
+km_kbp_cp *km::kbp::kmx::u16dup(km_kbp_cp *src) {
+  km_kbp_cp *dup = new km_kbp_cp[u16len(src) + 1];
+  memcpy(dup, src, (u16len(src) + 1) * 2);
```
```suggestion
  memcpy(dup, src, (u16len(src) + 1) * sizeof(km_kbp_cp));
```

mcdurdin commented on common/core/desktop/src/km_kbp_keyboard_api.cpp:
```diff
@@ -98,3 +98,27 @@ void km_kbp_keyboard_key_list_dispose(km_kbp_keyboard_key *key_list)
+
+void km_kbp_keyboard_imx_list_dispose(km_kbp_keyboard_imx *imx_list)
+{
+  km_kbp_keyboard_imx *imx_rule_it = imx_list;
```
```suggestion
  if(!imx_list) {
    return;
  }
  
  km_kbp_keyboard_imx *imx_rule_it = imx_list;
```

mcdurdin reviewed (approved):
LGTM!

MakaraSok commented:
All tests done in Windows 10 Pro 21H1 (VBox machine)

## SUITE_SIMPLIFIED_CHINESE:
   - **TEST_SIMPLIFIED_SINGLE: FAILED** the IMX window doesn't appear after pressing <kbd>o</kbd>. It seems like the keyboard (cs-pinyin) doesn't work at all. The same issue is experienced in LibreOffice.
   - **TEST_SIMPLIFIED_MULTIPLE: BLOCKED** It seems like the keyboard (cs-pinyin) doesn't work at all.
   - **TEST_SIMPLIFIED_BACKSPACE_1: BLOCKED** It seems like the keyboard (cs-pinyin) doesn't work at all.
   - **TEST_SIMPLIFIED_BACKSPACE_1: BLOCKED** It seems like the keyboard (cs-pinyin) doesn't work at all.

## SUITE_IMSAMPLE_KEYBOARD:

They two keys have been added and the value set to "0" like so:
<img width="300" src="https://user-images.githubusercontent.com/28331388/150686121-dff00517-9b9c-4245-b667-cf39af398235.png">

   - **TEST_IMSAMPLE_INPUT: FAILED** the IMX window doesn't appear after pressing <kbd>o</kbd>. It seems like the keyboard (cs-pinyin) doesn't work at all. The same issue is experienced in LibreOffice.
   - **TEST_IMSAMPLE_INPUT_CONT: BLOCKED**
   - **TEST_IMSAMPLE_BACKSPACE: BLOCKED**
   - **TEST_IMSAMPLE_BACKSPACE: BLOCKED**

## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW:
They two keys' value is set to "1" like so:
<img width="300" src="https://user-images.githubusercontent.com/28331388/150686242-48fc65c2-c519-4d74-832b-b297d5f7e547.png">
   - **TEST_IMSAMPLE_INPUT_IMW: FAILED** the IMX window doesn't appear after pressing <kbd>o</kbd>. It seems like the keyboard (cs-pinyin) doesn't work at all. The same issue is experienced in LibreOffice.
   - **TEST_IMSAMPLE_INPUT_IMW_CONT: BLOCKED**
   - **TEST_IMSAMPLE_BACKSPACE_IMW: BLOCKED**

rc-swag commented:
I will look into this straight away. However, just noting test `SUITE_IMSAMPLE_KEYBOARD` The IM window is not meant to show. 
In this mode IM window is not shown rather the context is update with a menu when either aeom is pressed.
For example when you  type the text will become h[1ɛ 2ɜ 3ə 4e 5ɘ] then is you type the number 3 the text displayed should now be hə

It is though in the third Suite `SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW`

MakaraSok commented:
> I will look into this straight away. However, just noting test `SUITE_IMSAMPLE_KEYBOARD` The IM window is not meant to show.

Everything is copied from the "User Test Results" above. Nothing has been changed, except the results of the test and comments.

Please double check the user test instructions provided.

MakaraSok commented:
## SUITE_SIMPLIFIED_CHINESE: 
* **TEST_SIMPLIFIED_SINGLE: PASSED** The IMX works and the expected character got output when tested in Notepad.
* **TEST_SIMPLIFIED_MULTIPLE: PASSED** The IMX works and the expected character got output when tested in Notepad.
* **TEST_SIMPLIFIED_BACKSPACE_1: PASSED** The IMX works and the expected character got output when tested in Notepad.
* **TEST_SIMPLIFIED_BACKSPACE_1: FAILED** when pressing the backspace twice for the second time, nothing is left in the line.

## SUITE_IMSAMPLE_KEYBOARD: 
* **TEST_IMSAMPLE_INPUT: FAILED** Keyman crashes when typing either "he" or "try".
Here is what's set in RegEdit: <br/>
<img width="300" src="https://user-images.githubusercontent.com/28331388/150947696-72c47225-c55c-440b-b1ac-87bf7b06f555.png"> <br/>
* **TEST_IMSAMPLE_INPUT_CONT: BLOCKED** Not testable because of the issue right above.
* **TEST_IMSAMPLE_BACKSPACE: BLOCKED** Not testable because of the issue right above.
* **TEST_IMSAMPLE_BACKSPACE: BLOCKED** Not testable because of the issue right above.

## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW: 
* **TEST_IMSAMPLE_INPUT_IMW: FAILED** Keyman crashes when tryping "try a". <br/>

https://user-images.githubusercontent.com/28331388/150952093-9c4feeb8-1668-4172-ab0c-861759522c9a.mov 

* **TEST_IMSAMPLE_INPUT_IMW_CONT: BLOCKED**
* **TEST_IMSAMPLE_BACKSPACE_IMW: BLOCKED**

MakaraSok commented:
@rc-swag Please refer me to the latest keyboard package for testing this. Thanks.

MakaraSok commented:
Also, please double check the test cases and ensure that there is no additional/duplicate cases in the sets/suites. Thanks.

rc-swag commented:
> Also, please double check the test cases and ensure that there is no additional/duplicate cases in the sets/suites. Thanks.

I tagged you in the test instructions were the Imsample.zip contains the latest keyboard. It was also in the conversation history but I removed it to avoid any confusion. 

I have been through the test instructions I renamed a duplicate name in the simplified Chinese backspace check.

MakaraSok commented:
## SUITE_SIMPLIFIED_CHINESE:

* **TEST_SIMPLIFIED_BACKSPACE_1: PASSED**
* **TEST_SIMPLIFIED_BACKSPACE_2: PASSED**

## SUITE_IMSAMPLE_KEYBOARD:

* **TEST_IMSAMPLE_INPUT: PASSED**
<img width="300" src="https://user-images.githubusercontent.com/28331388/151111481-f0a9ca93-9df6-4bc3-b6ed-8004169bd0c9.png">

* **TEST_IMSAMPLE_INPUT_CONT: PASSED**
<img width="300" src="https://user-images.githubusercontent.com/28331388/151111603-fd07b54e-6de9-4d50-9d04-7868cdeac95c.png">

* **TEST_IMSAMPLE_BACKSPACE: PASSED**
<img width="300" src="https://user-images.githubusercontent.com/28331388/151111722-c2529c6b-5d1a-4bb5-bc03-89074feba3ed.png">

* **TEST_IMSAMPLE_BACKSPACE_2: PASSED**

## SUITE_IMSAMPLE_KEYBOARD_IM_WINDOW:
Here is the settings I have in the regedit:
<img width="300" src="https://user-images.githubusercontent.com/28331388/151112527-b40de01c-9b14-4ce8-aae3-f9290bbba3c8.png">

* **TEST_IMSAMPLE_INPUT_IMW: PASSED**
* **TEST_IMSAMPLE_INPUT_IMW_CONT: PASSED**
* **TEST_IMSAMPLE_BACKSPACE_IMW: PASSED**

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 15.0.187-alpha](https://keyman.com/downloads/releases/alpha/15.0.187)
